torch/autograd/gradcheck.py

```diff
@@ -102,76 +114,132 @@ def get_stride(size):
             yield x_tensor, x_idx, d_idx
 
 
-def compute_gradient(fn, inputs, x, idx, delta, eps, is_mkldnn):
+def get_numerical_jacobian(fn, inputs, outputs=None, target=None, eps=1e-3,
+                           grad_out=1.0) -> List[Tuple[torch.Tensor]]:
+    """Computes the numerical jacobian for a given fn and inputs. Returns M * N jacobians
+    where M is the number of input tensors that require grad, and N is the number of output
+    float/complex tensors.
+
+    Args:
+        fn: the function to compute the jacobian for
+        inputs: inputs to `fn`
+        outputs: provide precomputed outputs to avoid one extra invocation of fn
+        target: the Tensors wrt whom Jacobians are calculated (default=`inputs`)
+        eps: the magnitude of the perturbation during finite differencing (default=`1e-3`)
+        grad_out: grad output value used to calculate gradients.
+
+    Returns:
+        M lists of N-tuples of jacobians
+
+    Note that `target` may not even be part of `input` to `fn`, so please be
+    **very careful** in this to not clone `target`.
+    """
+    jacobians: List[Tuple[torch.Tensor]] = []
+    if outputs is None:
+        outputs = _as_tuple(fn(inputs))
+    if target is None:
+        target = inputs
+    inp_indices = [i for i, a in enumerate(target) if is_tensor_like(a) and a.requires_grad]
+    for i, (inp, inp_idx) in enumerate(zip(iter_tensors(target, True), inp_indices)):
+        jacobians += [get_numerical_jacobian_helper(fn, inp, inp_idx, inputs, outputs, eps, grad_out)]
+    return jacobians
+
+
+def prepped_input(input, input_idx, entry, entry_idx):
+    # Prepares the inputs to be passed into the function while including the new modified input.
```

Comment:
The way its called is that instead of doing something like `f(tupled_inputs)`, we first wrap it in `f([prepped_input(a, i, entry, entry_idx) for i, a in enumerate(tupled_inputs))]`. `entry` is the particular input was is perturbed for the calculation (since we want to compute the numerical jacobian wrt to that particular input). `input` is the `a` that we happen to loop to. Depending on the layout of input, and whether `input_idx` is equal to `entry_idx` we may want to "prepare" it first before passing it back to `f`.

For example if we have some f taking a, b, where a is strided and b is mkldnn. We we need to compute derivatives wrt b, we need to 1) use the entry we modified instead of input because entry is a clone so changes in entry are not reflected in input 2) we should convert it back to mkldnn with .to_mkldnn because f expects b to be mkldnn. 

Conversely, we cannot always simply return entry, because in the strided case for example, entry is input.data, which is not part of the original graph, and this would result in a failure for the gradgrad case.

Reply:
I think what really threw me off here was looking at `input_idx` and `entry_idx` and wondering when they're different or the same. Reading the callsite of prepped_input helped that understanding.

What if we changed this function to:
```
prepped_input(input, maybe_perturbed_input)
```
?
`maybe_perturbed_input` is the perturbed input if it corresponds to this input, otherwise it is None. So we'd change `fn_out` to be:
```
    def fn_out():
        inp = tuple(prepped_input(input, x if i == input_idx else None) if is_tensor_like(input) else input
                    for i, input in enumerate(_as_tuple(inputs)))
        return tuple(a.clone() for a in _as_tuple(fn(*inp)))
```